### ISSUE TITLE
Ethernet padding to minimum 64byte frame size

### ISSUE THREAD

**Issue by maruu** [seladb/PcapPlusPlus#531]
Hi,

I am currently trying to send TCP SYN/ACKs using IPv4.

The minimum Ethernet frame length is 64bytes. When I use PcapPlusPlus to generate a Packet containing Ethernet / IPv4 / TCP I end up with a packet which is smaller than that.

![image](https://user-images.githubusercontent.com/2047922/93450126-e01de700-f8d5-11ea-86ec-0e0fce712572.png)

![image](https://user-images.githubusercontent.com/2047922/93450652-7fdb7500-f8d6-11ea-8e04-b1c6d4a64e01.png)

As you can see, the Ethernet layer already provides Padding - but includes the CRC in the padding, which leads to a packet that is too small.

Looking at the available tools I have, I see that you have already thought about parsing padding bytes at the end of a Packet by using the PacketTrailerLayer. Unfortunately, I cannot use this Layer to provide padding to my Ethernet frame since it cannot be created by the user.

I am using Ubuntu 18.04 and your raw socket implementation to send the frames on the loopback IF.

Can you please explain how I would be able to create said TCP SYN/ACK frame with additional padding so it fulfills the 64byte requirement?

Thanks and all the best,
maruu

**Comment by maruu** [comment]
After some additional research I think I have found an answer to my question: PcapPlusPlus is not responsible for Ethernet CRC calculation as well as for the addition of padding bytes. Since I am using a loopback adapter in connection with raw sockets this could lead to the problems that I am currently facing.. 

I now understand that this is a not a problem with PcapPlusPlus, but maybe you can think of a solution to my problem as I am sure that you are more experienced with problems like these?

Thanks in advance!

maruu

**Comment by maruu** [comment]
After further inspection of the issue I can conclude that on a Loopback interface you don't need an Ethernet Frame Checksum. The problem with padding still persists though. A workaround that I am currently using is that I am trying to add options so that I get above the 64byte minimum frame length limit. Not preferable, but I think that it will work, as I have tested the reference implementation using netcat.

Thanks for (possibly) reading and sorry for taking your time :-)

**Comment by seladb** [comment]
I'm sorry for the late response. As you mentioned PcapPlusPlus currently doesn't add or handle CRC in the end of an Ethernet frame. As a workaround you can calculate the CRC yourself and add it to the end of the packet as a `PayloadLayer`. If this feature is needed we can consider adding it.

**Comment by maruu** [comment]
I have found that the problem is not with the CRC but more with the length that is not being padded by the components involved when sending over a loopback interface. It would be great if one could add padding themselves, by lets say adding the PacketTrailerLayer with a certain length. But I will investigate further. For now, I have found a workaround, as described before.

Thanks for your answer and time.
